File: senmarket-frontend/src/lib/api/services/quota.service.ts
// src/lib/api/services/quota.service.ts - VERSION CORRIGÉE ALIGNÉE BACKEND
import { apiClient } from '../client';
import { ApiResponse } from '../types';

// ============================================
// TYPES ALIGNÉS AVEC VOTRE BACKEND GO
// ============================================

export interface QuotaStatus {
  current_phase: string;
  is_launch_active: boolean;
  launch_end_date?: string;
  days_until_launch_end?: number;
  currency: string;
  standard_price: number;
  unlimited_free: boolean;
  remaining_free: number;
  monthly_limit?: number;
  used_this_month?: number;
  paid_this_month?: number;
  total_this_month?: number;
  can_create_free: boolean;
  quota_progress?: number;
  reset_date?: string;
  days_until_reset?: number;
  period?: string;
  phase_name: string;
  phase_description: string;
  message: string;
}

export interface EligibilityCheck {
  can_create_free: boolean;
  requires_payment: boolean;
  current_phase: string;
  standard_price: number;
  currency: string;
  reason?: string;
  quota_reset_date?: string;
  days_until_reset?: number;
  used_this_month?: number;
  limit_this_month?: number;
  premium_options?: {
    boost_price: number;
    featured_price: number;
    pack_5_price: number;
    pack_10_price: number;
    pack_5_discount: number;
    pack_10_discount: number;
  };
}

export interface QuotaSummary {
  can_create_free: boolean;
  remaining_free: number;
  unlimited_free: boolean;
  requires_payment: boolean;
  message: string;
  current_phase: string;
  price_per_listing?: number;
  currency?: string;
}

export interface QuotaHistory {
  months_requested: number;
  periods_found: number;
  history: Array<{
    period: string;
    month: number;
    year: number;
    free_used: number;
    free_limit: number;
    free_remaining: number;
    paid_listings: number;
    total_listings: number;
    can_create_free: boolean;
    progress_percent: number;
    is_current_month: boolean;
  }>;
  summary?: {
    total_free_used: number;
    total_paid: number;
    total_listings: number;
  };
}

export interface CurrentPhase {
  current_phase: string;
  is_launch_active: boolean;
  launch_end_date?: string;
  days_until_launch_end?: number;
  credit_system_active: boolean;
  paid_system_active: boolean;
  monthly_free_limit: number;
  standard_price: number;
  currency: string;
  phase_name: string;
  phase_description: string;
  benefits: string[];
}

export interface PlatformStats {
  current_phase: string;
  is_launch_active: boolean;
  days_until_launch_end?: number;
  current_month: number;
  current_year: number;
  total_users: number;
  total_listings: number;
  active_listings: number;
  free_listings_this_month: number;
  paid_listings_this_month: number;
  total_listings_this_month: number;
  revenue_this_month: number;
  active_users_this_month: number;
}

// ============================================
// SERVICE CLASS ALIGNÉ AVEC VOS ENDPOINTS
// ============================================

export class QuotaService {
  
  // ✅ Endpoint exact : /quota/status (QuotaHandler.GetQuotaStatus)
  async getQuotaStatus(): Promise<QuotaStatus> {
    const response = await apiClient.get<ApiResponse<QuotaStatus>>('/quota/status');
    return response.data.data;
  }

  // ✅ Endpoint exact : /quota/check (QuotaHandler.CheckEligibility)
  async checkEligibility(): Promise<EligibilityCheck> {
    const response = await apiClient.get<ApiResponse<EligibilityCheck>>('/quota/check');
    return response.data.data;
  }

  // ✅ Endpoint exact : /quota/summary (QuotaHandler.GetQuotaSummary)
  async getQuotaSummary(): Promise<QuotaSummary> {
    const response = await apiClient.get<ApiResponse<QuotaSummary>>('/quota/summary');
    return response.data.data;
  }

  // ✅ Endpoint exact : /quota/history (QuotaHandler.GetQuotaHistory)
  async getQuotaHistory(months: number = 6): Promise<QuotaHistory> {
    const response = await apiClient.get<ApiResponse<QuotaHistory>>('/quota/history', {
      params: { months }
    });
    return response.data.data;
  }

  // ✅ Endpoint exact : /quota/current-phase (QuotaHandler.GetCurrentPhase)
  async getCurrentPhase(): Promise<CurrentPhase> {
    const response = await apiClient.get<ApiResponse<CurrentPhase>>('/quota/current-phase');
    return response.data.data;
  }

  // ✅ Endpoint exact : /quota/platform-stats (QuotaHandler.GetPlatformStats)  
  async getPlatformStats(): Promise<PlatformStats> {
    const response = await apiClient.get<ApiResponse<PlatformStats>>('/quota/platform-stats');
    return response.data.data;
  }

  // ✅ Endpoint exact : /quota/update-phase (QuotaHandler.UpdateUserPhase)
  async updateUserPhase(): Promise<QuotaStatus> {
    const response = await apiClient.post<ApiResponse<QuotaStatus>>('/quota/update-phase');
    return response.data.data;
  }

  // ============================================
  // MÉTHODES UTILITAIRES POUR LE FRONTEND
  // ============================================

  // Vérifier si l'utilisateur est en phase de lancement gratuit
  isInLaunchPhase(status: QuotaStatus): boolean {
    return status.current_phase === 'launch' && status.is_launch_active;
  }

  // Vérifier si l'utilisateur peut créer une annonce gratuite
  canCreateFree(status: QuotaStatus): boolean {
    return status.can_create_free;
  }

  // Obtenir le message d'état adapté
  getStatusMessage(status: QuotaStatus): string {
    return status.message;
  }

  // Obtenir la couleur du statut pour l'UI
  getStatusColor(status: QuotaStatus): 'green' | 'yellow' | 'red' {
    if (this.isInLaunchPhase(status)) {
      return 'green'; // Phase gratuite
    }
    
    if (status.can_create_free) {
      return 'green'; // Peut encore créer gratuitement
    }
    
    if (status.remaining_free === 0) {
      return 'red'; // Quota épuisé
    }
    
    return 'yellow'; // Quota partiel
  }

  // Calculer le pourcentage d'utilisation du quota
  getQuotaUsagePercent(status: QuotaStatus): number {
    if (status.unlimited_free || !status.monthly_limit) {
      return 0;
    }
    
    return Math.round((status.used_this_month || 0) / status.monthly_limit * 100);
  }

  // Formater le prix pour l'affichage (FCFA au Sénégal)
  formatPrice(price: number, currency: string = 'FCFA'): string {
    return `${Math.round(price).toLocaleString('fr-FR')} ${currency}`;
  }

  // Calculer les jours restants jusqu'à la fin de la phase de lancement
  getDaysUntilLaunchEnd(launchEndDate?: string): number {
    if (!launchEndDate) return 0;
    
    const endDate = new Date(launchEndDate);
    const now = new Date();
    const diffTime = endDate.getTime() - now.getTime();
    return Math.max(0, Math.ceil(diffTime / (1000 * 60 * 60 * 24)));
  }

  // Vérifier si la phase de lancement se termine bientôt (moins de 7 jours)
  isLaunchEndingSoon(status: QuotaStatus): boolean {
    return this.isInLaunchPhase(status) && (status.days_until_launch_end || 0) <= 7;
  }

  // Obtenir le texte d'urgence pour la fin de phase
  getUrgencyMessage(status: QuotaStatus): string | null {
    if (!this.isInLaunchPhase(status)) {
      return null;
    }
    
    const days = status.days_until_launch_end || 0;
    
    if (days <= 1) {
      return '⚡ Plus que quelques heures pour profiter des annonces gratuites !';
    }
    
    if (days <= 3) {
      return `⚡ Plus que ${days} jours pour profiter des annonces gratuites !`;
    }
    
    if (days <= 7) {
      return `⏰ Plus que ${days} jours pour profiter des annonces gratuites.`;
    }
    
    return null;
  }

  // Obtenir les recommandations d'action pour l'utilisateur
  getActionRecommendations(status: QuotaStatus): Array<{
    action: string;
    priority: 'high' | 'medium' | 'low';
    message: string;
  }> {
    const recommendations = [];
    
    // Phase de lancement qui se termine
    if (this.isLaunchEndingSoon(status)) {
      recommendations.push({
        action: 'create_listings_now',
        priority: 'high' as const,
        message: 'Créez vos annonces maintenant pendant que c\'est gratuit !'
      });
    }
    
    // Quota épuisé en phase de crédits
    if (status.current_phase === 'credit_system' && !status.can_create_free) {
      recommendations.push({
        action: 'wait_or_pay',
        priority: 'medium' as const,
        message: `Attendez ${status.days_until_reset} jours ou payez ${this.formatPrice(status.standard_price, status.currency)} pour publier.`
      });
    }
    
    // Quota bientôt épuisé
    if (status.remaining_free <= 1 && status.remaining_free > 0) {
      recommendations.push({
        action: 'use_remaining',
        priority: 'medium' as const,
        message: `Plus qu'${status.remaining_free} annonce gratuite ce mois !`
      });
    }
    
    return recommendations;
  }

  // ✅ NOUVELLES MÉTHODES ALIGNÉES AVEC VOTRE BACKEND

  // Obtenir un résumé rapide pour l'UI (cache-friendly)
  async getQuickStatus(): Promise<{
    canCreateFree: boolean;
    message: string;
    phase: string;
    priceRequired?: number;
  }> {
    try {
      const summary = await this.getQuotaSummary();
      return {
        canCreateFree: summary.can_create_free,
        message: summary.message,
        phase: summary.current_phase,
        priceRequired: summary.requires_payment ? summary.price_per_listing : undefined
      };
    } catch (error) {
      // Fallback en cas d'erreur
      return {
        canCreateFree: false,
        message: 'Impossible de vérifier le statut',
        phase: 'unknown'
      };
    }
  }

  // Vérifier si l'utilisateur est dans une phase payante
  isInPaidPhase(status: QuotaStatus): boolean {
    return status.current_phase === 'paid_system' || 
           (status.current_phase === 'credit_system' && !status.can_create_free);
  }

  // Obtenir le nombre d'annonces restantes avant paiement
  getRemainingFreeListings(status: QuotaStatus): number {
    if (status.unlimited_free) return -1; // Illimité
    return Math.max(0, status.remaining_free);
  }

  // Formatter les informations de quota pour l'affichage
  formatQuotaInfo(status: QuotaStatus): {
    title: string;
    description: string;
    color: 'green' | 'yellow' | 'red';
    icon: string;
  } {
    if (this.isInLaunchPhase(status)) {
      return {
        title: '🎉 Période de lancement',
        description: 'Annonces gratuites illimitées !',
        color: 'green',
        icon: '🚀'
      };
    }

    if (status.can_create_free) {
      const remaining = this.getRemainingFreeListings(status);
      return {
        title: `${remaining} annonce${remaining > 1 ? 's' : ''} gratuite${remaining > 1 ? 's' : ''}`,
        description: `Il vous reste ${remaining} annonce${remaining > 1 ? 's' : ''} gratuite${remaining > 1 ? 's' : ''} ce mois`,
        color: remaining > 1 ? 'green' : 'yellow',
        icon: '✅'
      };
    }

    return {
      title: 'Quota épuisé',
      description: `${this.formatPrice(status.standard_price, status.currency)} par annonce`,
      color: 'red',
      icon: '💳'
    };
  }
}

// ============================================
// INSTANCE SINGLETON
// ============================================
export const quotaService = new QuotaService();
export default quotaService;

File: senmarket-backend-backup-20250628-153135/internal/services/quota_service.go
// internal/services/quota_service.go
package services

import (
	"errors"
	"fmt"
	"time"

	"senmarket/internal/models"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

var (
	ErrNoFreeListingsLeft = errors.New("quota d'annonces gratuites épuisé pour ce mois")
	ErrInvalidUser        = errors.New("utilisateur invalide")
	ErrConfigNotFound     = errors.New("configuration de prix non trouvée")
)

type QuotaService struct {
	db *gorm.DB
}

func NewQuotaService(db *gorm.DB) *QuotaService {
	return &QuotaService{db: db}
}

// GetGlobalConfig récupère la configuration globale de tarification
func (s *QuotaService) GetGlobalConfig() (*models.PricingConfig, error) {
	return models.GetOrCreateGlobalConfig(s.db)
}

// IsLaunchPhaseActive vérifie si la phase de lancement gratuit est active
func (s *QuotaService) IsLaunchPhaseActive() (bool, error) {
	config, err := s.GetGlobalConfig()
	if err != nil {
		return false, err
	}
	
	return config.IsFreeLaunchActive(), nil
}

// GetOrCreateMonthlyQuota récupère ou crée le quota mensuel d'un utilisateur
func (s *QuotaService) GetOrCreateMonthlyQuota(userID uuid.UUID) (*models.ListingQuota, error) {
	now := time.Now()
	month := int(now.Month())
	year := now.Year()
	
	return models.GetOrCreateQuotaForPeriod(s.db, userID, month, year)
}

// CanCreateFreeListing vérifie si l'utilisateur peut créer une annonce gratuite
func (s *QuotaService) CanCreateFreeListing(userID uuid.UUID) (bool, *models.ListingQuota, error) {
	// Vérifier la phase globale d'abord
	isLaunchActive, err := s.IsLaunchPhaseActive()
	if err != nil {
		return false, nil, err
	}
	
	// Phase 1: Lancement gratuit - illimité
	if isLaunchActive {
		return true, nil, nil
	}
	
	// Phase 2+: Système de quotas mensuels
	quota, err := s.GetOrCreateMonthlyQuota(userID)
	if err != nil {
		return false, nil, err
	}
	
	return quota.CanCreateFreeListing(), quota, nil
}

// ConsumeFreeListing consomme une annonce gratuite
func (s *QuotaService) ConsumeFreeListing(userID uuid.UUID) error {
	// Vérifier la phase globale
	isLaunchActive, err := s.IsLaunchPhaseActive()
	if err != nil {
		return err
	}
	
	// Phase 1: Pas de consommation nécessaire (illimité)
	if isLaunchActive {
		return nil
	}
	
	// Phase 2+: Consommer le quota mensuel
	quota, err := s.GetOrCreateMonthlyQuota(userID)
	if err != nil {
		return err
	}
	
	if !quota.CanCreateFreeListing() {
		return ErrNoFreeListingsLeft
	}
	
	return quota.ConsumeFreeListing(s.db)
}

// AddPaidListing ajoute une annonce payée au compteur
func (s *QuotaService) AddPaidListing(userID uuid.UUID) error {
	// Toujours compter les annonces payées, même en phase de lancement
	quota, err := s.GetOrCreateMonthlyQuota(userID)
	if err != nil {
		return err
	}
	
	return quota.AddPaidListing(s.db)
}

// GetUserQuotaStatus retourne le statut détaillé du quota utilisateur
func (s *QuotaService) GetUserQuotaStatus(userID uuid.UUID) (map[string]interface{}, error) {
	config, err := s.GetGlobalConfig()
	if err != nil {
		return nil, err
	}
	
	status := map[string]interface{}{
		"current_phase":        config.GetCurrentPhase(),
		"is_launch_active":     config.IsFreeLaunchActive(),
		"launch_end_date":      config.LaunchPhaseEndDate,
		"days_until_launch_end": config.GetDaysUntilLaunchEnd(),
		"currency":             config.Currency,
		"standard_price":       config.StandardListingPrice,
	}
	
	// Phase 1: Lancement gratuit
	if config.IsFreeLaunchActive() {
		status["unlimited_free"] = true
		status["remaining_free"] = -1 // -1 = illimité
		status["phase_name"] = "Lancement gratuit"
		status["phase_description"] = "Publication illimitée et gratuite"
		status["message"] = config.GetPhaseMessage(-1)
		return status, nil
	}
	
	// Phase 2+: Quotas mensuels
	quota, err := s.GetOrCreateMonthlyQuota(userID)
	if err != nil {
		return nil, err
	}
	
	remaining := quota.RemainingFreeListings()
	
	status["unlimited_free"] = false
	status["monthly_limit"] = quota.FreeListingsLimit
	status["used_this_month"] = quota.FreeListingsUsed
	status["remaining_free"] = remaining
	status["paid_this_month"] = quota.PaidListings
	status["total_this_month"] = quota.FreeListingsUsed + quota.PaidListings
	status["can_create_free"] = quota.CanCreateFreeListing()
	status["quota_progress"] = quota.GetProgress()
	status["reset_date"] = quota.NextResetDate()
	status["days_until_reset"] = quota.DaysUntilReset()
	status["period"] = quota.GetPeriodString()
	
	// Messages adaptés à la phase
	if config.GetCurrentPhase() == "credit_system" {
		status["phase_name"] = "Système de crédits"
		status["phase_description"] = fmt.Sprintf("%d annonces gratuites par mois", quota.FreeListingsLimit)
	} else {
		status["phase_name"] = "Système payant"
		status["phase_description"] = fmt.Sprintf("%.0f FCFA par annonce", config.StandardListingPrice)
	}
	
	status["message"] = config.GetPhaseMessage(remaining)
	
	return status, nil
}

// GetUserQuotaHistory récupère l'historique des quotas d'un utilisateur
func (s *QuotaService) GetUserQuotaHistory(userID uuid.UUID, months int) ([]models.ListingQuota, error) {
	if months <= 0 {
		months = 6 // Par défaut, 6 derniers mois
	}
	
	return models.GetUserQuotaHistory(s.db, userID, months)
}

// CheckListingEligibility vérifie l'éligibilité complète pour créer une annonce
func (s *QuotaService) CheckListingEligibility(userID uuid.UUID) (map[string]interface{}, error) {
	canCreateFree, quota, err := s.CanCreateFreeListing(userID)
	if err != nil {
		return nil, err
	}
	
	config, err := s.GetGlobalConfig()
	if err != nil {
		return nil, err
	}
	
	result := map[string]interface{}{
		"can_create_free":   canCreateFree,
		"requires_payment":  !canCreateFree,
		"current_phase":     config.GetCurrentPhase(),
		"standard_price":    config.StandardListingPrice,
		"currency":          config.Currency,
	}
	
	if !canCreateFree && quota != nil {
		result["reason"] = fmt.Sprintf(
			"Vous avez utilisé vos %d annonces gratuites ce mois-ci", 
			quota.FreeListingsLimit,
		)
		result["quota_reset_date"] = quota.NextResetDate()
		result["days_until_reset"] = quota.DaysUntilReset()
		result["used_this_month"] = quota.FreeListingsUsed
		result["limit_this_month"] = quota.FreeListingsLimit
	}
	
	// Informations sur les options premium (pour phase 3)
	if config.PaidSystemActive {
		result["premium_options"] = map[string]interface{}{
			"boost_price":    config.PremiumBoostPrice,
			"featured_price": config.FeaturedColorPrice,
			"pack_5_price":   config.Pack5ListingsPrice,
			"pack_10_price":  config.Pack10ListingsPrice,
			"pack_5_discount": config.Pack5Discount,
			"pack_10_discount": config.Pack10Discount,
		}
	}
	
	return result, nil
}

// GetPlatformStats retourne des statistiques globales de la plateforme
func (s *QuotaService) GetPlatformStats() (map[string]interface{}, error) {
	config, err := s.GetGlobalConfig()
	if err != nil {
		return nil, err
	}
	
	now := time.Now()
	currentMonth := int(now.Month())
	currentYear := now.Year()
	
	stats := map[string]interface{}{
		"current_phase":        config.GetCurrentPhase(),
		"is_launch_active":     config.IsFreeLaunchActive(),
		"days_until_launch_end": config.GetDaysUntilLaunchEnd(),
		"current_month":        currentMonth,
		"current_year":         currentYear,
	}
	
	// Compter les utilisateurs totaux
	var totalUsers int64
	s.db.Model(&models.User{}).Count(&totalUsers)
	stats["total_users"] = totalUsers
	
	// Compter les annonces totales
	var totalListings int64
	s.db.Model(&models.Listing{}).Count(&totalListings)
	stats["total_listings"] = totalListings
	
	// Compter les annonces actives
	var activeListings int64
	s.db.Model(&models.Listing{}).Where("status = ?", "active").Count(&activeListings)
	stats["active_listings"] = activeListings
	
	// Stats du mois en cours
	var monthlyQuotas []models.ListingQuota
	s.db.Where("month = ? AND year = ?", currentMonth, currentYear).Find(&monthlyQuotas)
	
	var freeListingsThisMonth, paidListingsThisMonth int64
	for _, quota := range monthlyQuotas {
		freeListingsThisMonth += int64(quota.FreeListingsUsed)
		paidListingsThisMonth += int64(quota.PaidListings)
	}
	
	stats["free_listings_this_month"] = freeListingsThisMonth
	stats["paid_listings_this_month"] = paidListingsThisMonth
	stats["total_listings_this_month"] = freeListingsThisMonth + paidListingsThisMonth
	
	// Calcul du revenu estimé (phase 2+)
	var revenueThisMonth float64
	if !config.IsFreeLaunchActive() {
		revenueThisMonth = float64(paidListingsThisMonth) * config.StandardListingPrice
	}
	stats["revenue_this_month"] = revenueThisMonth
	
	// Utilisateurs actifs (avec au moins une annonce ce mois)
	var activeUsers int64
	s.db.Model(&models.ListingQuota{}).
		Where("month = ? AND year = ? AND (free_listings_used > 0 OR paid_listings > 0)", 
			currentMonth, currentYear).
		Count(&activeUsers)
	stats["active_users_this_month"] = activeUsers
	
	return stats, nil
}

// TransitionUserToNextPhase fait passer un utilisateur à la phase suivante
func (s *QuotaService) TransitionUserToNextPhase(userID uuid.UUID) error {
	var user models.User
	if err := s.db.First(&user, userID).Error; err != nil {
		return fmt.Errorf("utilisateur non trouvé: %w", err)
	}
	
	return user.TransitionToNextPhase(s.db)
}

// TransitionGlobalToNextPhase fait passer la plateforme à la phase suivante
func (s *QuotaService) TransitionGlobalToNextPhase(adminUserID uuid.UUID) error {
	config, err := s.GetGlobalConfig()
	if err != nil {
		return err
	}
	
	return config.TransitionToNextPhase(s.db, &adminUserID)
}

// ExtendLaunchPhase prolonge la phase de lancement
func (s *QuotaService) ExtendLaunchPhase(newEndDate time.Time, adminUserID uuid.UUID) error {
	config, err := s.GetGlobalConfig()
	if err != nil {
		return err
	}
	
	return config.ExtendLaunchPhase(s.db, newEndDate, &adminUserID)
}

// UpdateGlobalPrices met à jour les tarifs globaux
func (s *QuotaService) UpdateGlobalPrices(newPrices map[string]float64, adminUserID uuid.UUID) error {
	config, err := s.GetGlobalConfig()
	if err != nil {
		return err
	}
	
	return config.UpdatePrices(s.db, newPrices, &adminUserID)
}

// CleanupOldQuotas nettoie les anciens quotas (à appeler périodiquement)
func (s *QuotaService) CleanupOldQuotas() error {
	return models.CleanupOldQuotas(s.db)
}

// GetQuotaSummaryForUser retourne un résumé simple du quota utilisateur
func (s *QuotaService) GetQuotaSummaryForUser(userID uuid.UUID) (map[string]interface{}, error) {
	status, err := s.GetUserQuotaStatus(userID)
	if err != nil {
		return nil, err
	}
	
	// Version simplifiée pour l'API frontend
	summary := map[string]interface{}{
		"can_create_free":    status["can_create_free"],
		"remaining_free":     status["remaining_free"],
		"unlimited_free":     status["unlimited_free"],
		"requires_payment":   !status["can_create_free"].(bool),
		"message":           status["message"],
		"current_phase":     status["current_phase"],
	}
	
	// Ajouter les infos de prix si nécessaire
	if !status["can_create_free"].(bool) {
		summary["price_per_listing"] = status["standard_price"]
		summary["currency"] = status["currency"]
	}
	
	return summary, nil
}